### DISCUSSION TITLE
MooseFS Distributed Metadata Server (Feature)

### DISCUSSION THREAD

**Discussion by anon314159** [moosefs/moosefs#608] [Ideas]
MooseFS Developers, 

What is the technical feasibility of implementing a distributed metadata (active-active) setup with MooseFS?  Conceptionally, it would mirror Ceph's design (i.e. distributed-replicated metadata servers) leveraging a quorum based system. The current design strives for much lower latencies and faster lookup times than other distributed systems like Ceph, GlusterFS, BeeGFS, Lustre, etc. but can easily hit a performance wall when subjected to a large number of clients even on pretty high-end hardware (high IPC rated processors based on Passmarks benchmarks). 

A practical example of this would be in the HPC space with say thousands of MooseFS clients accessing thousands of files concurrently (ex. Slurm + MooseFS).  I believe properly designed and spec'ed out chunk servers with a decent backend network can easily deliver the proper amount of I/O but are easily bottlenecked by the master servers ability to serve out these requests.  This isn't a critic of MooseFS's design, as I understand the rational behind its implementation and prefer a faster user experience when interacting with any distributed file system.  

I come from the POV of a long-time user of GlusterFS and understand all to well the pain of interacting with systems that lack centralized metadata. Given the ever increasing demand for ML/AI,  HPC, and big-data workloads, I believe a distributed master server would greatly increase MooseFS's reach outside of just general purpose file storage. I digress, how feasible would it be to implement either an active-active metadata server setup or better yet add multicore/multithreading support for the mfsmaster process?

**Comment by chogata** [comment]
This is something that we talk about internally from time to time. Basically, the current design has 2 advantages: it's fast and it's safe. Multi-threaded systems, no matter how well coded and tested, are always more error prone, because of the unpredictability of threads' concurrent operations. And you can't lock all of your code behind a critical section, because that's basically a single threaded program ;)
Multi-machine: well, that has its own challenges and while we could probably keep the same level of safety, the impact on performance would be big.
Having said that, we have something on our roadmap that will help: we intend to move all network operations to a separate sub-process, and not a single one, either. An analysis of current master's work shows us that it looks like it spends more than 50% of its time doing "network stuff", and even more so with many clients connected. So this will certainly help in bigger instances with many clients.

**Reply by ptempier** [reply]
Maybe allow it only for the "easy" case ?
If a file is somehow readonly :  in trash, in RO snapshot, in a worm storage.

Or create a special thing like auto-reWORM every week.
Like your packages repository you guarantee that it can't be deleted for one week.
And at the end of the week, either you renew that guaranty, or set the policy to delete it or allow writes again.

I guess it works well for for example for IA training dataset that are not changed, backups, archives, etc ...

**Reply by ptempier** [reply]
Thinking another way is to  basically zone the metadata.
You have /mnt/mfs/siteA  /mnt/mfs/siteB
Maybe there can be a metadataserver that's master for siteA and another for siteB
And the other is the slave of it for that part of the tree.

That would reduce latency for peoples in different sites,
and half the number of requests for a particular server.

**Reply by chogata** [reply]
In the setup you are describing, with 2 zones, it looks like you could just use 2 MooseFS instances. Use master from site A as metalogger for site B and vice versa. Any independent data set can just have its own MooseFS instance.

**Reply by ptempier** [reply]
Yes, almost the same, hence why i suggested it as it didn't seem too difficult to implement.
My undertsanding is that a mapping of:  folder-> /master/meta need to be send to the client,
It would not be very long and  then almost never need to be updated.
Then the client can speak directly to the proper master for that part of the tree.

If the client want to move a file and that implies a master change for theses files,
i think the data path is a bit long, but it shouldnt  happen very often.
Also it would just be metadata/ownership info moved around, not data.
client->(ask move)->newmaster->(ask transfert ownership)->oldmaster->(notify ownership)->chunckserver->(done)->oldmaster->(done)->newmaster->(done)->client

A few minor differences  with 2 instances:
2 chunkserver instances can't share the same chunck folder
One instance may not have the notion that a file is already chunked/crcred,etc when copied over its instance
A unic file hierarchy can't be done with 2 instances.
From a user pov, if he has 4 sites, he doesnt have to multiply by 4 the number of instances.

**Reply by ptempier** [reply]
I did some benchmark todays using the grid5000 cluster using io500.
(3 sites, 1 master, 6 chunks per site , 2 clients per site)

If the client is not in the same datacenter than the master :
Data manipulation is ~7x slower ( I believe that's due to the intersite bandwith )
Metadata manipulation is ~700x slower ( I believe that's due to the intersite latency )
( That's a shortened explanation from ~75 metrics )

Using the /siteA /siteB thing would help in that case.
It would still leave the issue with moving data from a folder manager by one master to one managed by a different master.

**Reply by anon314159** [reply]
> Maybe allow it only for the "easy" case ? If a file is somehow readonly : in trash, in RO snapshot, in a worm storage.
> 
> Or create a special thing like auto-reWORM every week. Like your packages repository you guarantee that it can't be deleted for one week. And at the end of the week, either you renew that guaranty, or set the policy to delete it or allow writes again.
> 
> I guess it works well for for example for IA training dataset that are not changed, backups, archives, etc ...

Thanks for the excellent explanation.  Based on the source code and documentation for BeeGFS and Ceph, they managed to implement "reasonably safe" and performant distributed metadata systems where read operations can be completed from any metadata server in the cluster but writes are always atomic, synchronized via quarum, and implemented with a strong consistency model.  For scientific purposes I have benchmarked both Ceph and MooseFS on the same exact hardware and hit IOP/s limits pretty quickly with single metadata servers on both (FIO, Sysbench, and Slurm) but could easily scale past the issue by simply adding another metadata with Ceph, the same cannot be said with MooseFS.  

Essentially, I am stuck and can no longer scale performance wise.  For most users this isn't a problem but HPC throws the baby into the bath water with a single master server.  Honestly, I genuinely hate Ceph, its bloated code base, management overhead and overly complex design.  I digress, in terms of data storage and distributed systems, I always leverage the CAP theorem and consistency models when determining each particular use case.  What consistency model does MooseFS use for metadata and chunk servers?  Also, would using mfsbdev or sparse files mapped as loopback devices and then served out via NFS/iSCSI targets help resolve the lots of small files overhead and reduce the metadata workload?  I suspect this would greatly reduce the number of metadata transactions as we are no longer reading/writing individual files that requires the master server track this information, instead blocks of data which in turn, moves the bottleneck to the chunk servers (ideal outcome anyway).

**Reply by ptempier** [reply]
Yes, one of the advantages of moosefs is that its somehow very simple,
i found it resilient to server/controller/disk loss, generally able to repair itself.
Maybe i need to do some test on file corruption, but its a pita to test for.
the performance is not unreasonable, but not sky high either specially for small files.
But you pay the simplicity by having no way to scale metadata query/updates,
hence why i was trying to propose some enhancements.
(zoned data for multimaster and using loggers for ro data)

Using mfsbdev really helps; but basically you trade :
- the ability to have different clients updating many files, but not that fast; 
- for the limitation of having a single client updating many files but way faster.

It works like the loopback via nfs you speak about (that you can already do anywhere). 
It's a general problem for distributed and even remote shares.
But moosefs has a dedicated system to do that.

Really i suggest you bench moosefs a bit. (to reset during mfsbdev benchs, destroy/create new devices, don't truncate)
There's various bench tools : you could code one, use fio, or use io500 (which is designed for hpc)
If you have for example 24 disks enclosures, you can do 4 raid0 with 6 drives (moosefs will do the raid1/5).
For mechanical drives, this trades a useful speed boost for longer rebuild operations.

**Reply by anon314159** [reply]
> Yes, one of the advantages of moosefs is that its somehow very simple, i found it resilient to server/controller/disk loss, generally able to repair itself. Maybe i need to do some test on file corruption, but its a pita to test for. the performance is not unreasonable, but not sky high either specially for small files. But you pay the simplicity by having no way to scale metadata query/updates, hence why i was trying to propose some enhancements. (zoned data for multimaster and using loggers for ro data)
> 
> Using mfsbdev really helps; but basically you trade :
> 
> * the ability to have different clients updating many files, but not that fast;
> * for the limitation of having a single client updating many files but way faster.
> 
> It works like the loopback via nfs you speak about (that you can already do anywhere). It's a general problem for distributed and even remote shares. But moosefs has a dedicated system to do that.
> 
> Really i suggest you bench moosefs a bit. (to reset during mfsbdev benchs, destroy/create new devices, don't truncate) There's various bench tools : you could code one, use fio, or use io500 (which is designed for hpc) If you have for example 24 disks enclosures, you can do 4 raid0 with 6 drives (moosefs will do the raid1/5). For mechanical drives, this trades a useful speed boost for longer rebuild operations.

Thanks for the information. I empathize completely with keeping the design as simple as possible due to the possible introduction of unforeseen bugs or corruption.  Regarding benchmarking, I have never been I/O bound with MooseFS! And I have at least eight different clusters in a variety of configurations spanning from nvme PCI Express storage to traditional SAS hdds ranging simple clustered raid 5/6s to just jbod and all them perform very well in terms of throughput and bandwidth. However, the glaring issue is scalability with metadata operations and while raid and using loop devices via sparse files can greatly help reduce the overhead and improve latency, I don't believe it entirely resolves the metadata performance under heavy load. It seems as though MooseFS' is experiencing the very same problem ceph pre-luminous had and that is it could only support one active metadata server. Meaning, all metadata operations (directory listings, file lookups, permission checks, etc.) funneled through that one MDS. It was a major bottleneck for scalability.  How they were were able to overcome this problem involved the use of dynamic directory subtree partitioning (very similar to your zone concept).  

Conceptually it goes something like this (my interpretation based on the source code, so take it with a grain of salt lol):

The file system’s directory tree is divided dynamically into "subtrees."

Each subtree (i.e., a directory and all its descendants) is assigned to one MDS.

The MDS servers coordinate among themselves to balance the load by migrating subtrees between each other based on real-time access patterns.

Migration: A "hot" subtree (high I/O) can be moved from a busy MDS to a less busy one to balance the load.

Load-based and access-pattern based splitting: if a subtree becomes too large, it can be automatically split into smaller subtrees and distributed across MDS nodes.

Each MDS is simply responsible for caching and partitioning this information, whereas the actual metadata is backed by disks (or in the base of MooseFS, the MDS's memory) across all the MDS's.

This is kind of already done with the single master server service, since it periodically flushes metadata information from memory back to disk and the metalogger too.  It seems the only difference is the metalogger simply copies the contents of /var/lib/mfs from the master to its own instance every hour or so.  

I think the most difficult part involves coordination between multiple Master servers, running concurrently and ensuring that there's no data hazard issues.  Either way, this is just some food for thought and take it with a grain of salt.

All of cephfs write and create operations conducted by the MDS are atomic and follow a strong data consistency model. So technically speaking, implementing something similar on top of MooseFS wouldn't violate its existing data consistency model.  As both Ceph and MooseFS implement the same models for metadata.  Essentially, it is a lessons learned from one file system to another and not having to reinvent the wheel or add excessive amounts of complexity or technical debt.  

But in the end, I still agree with your philosophy. If it ain't broke, don't fix it.  MooseFS's performance issues are largely edge cases for now (i.e. HPC), but it is still in fact broke and does require fixing.  Aside from that, generally speaking, I've had a difficult time attempting to, under real world scenarios, hit a metadata performance wall with MFS assuming the master server has appropriately spec'ed out hardware.  Trust me, I've had a very fun time explaining the lots of files problem with my computer scientists, who think it's a good idea to dump hundreds of millions of files in a single directory and wonder why it takes a few minutes to list its contents lol.  This adds a bit context so that we are both on the same page use case wise.  A lot of these issues are user driven bad practices but some of them are simple inherent design flaws within distributed systems not exclusive to just MooseFS.  Interestingly enough, all of your recommendations you've listed above have been attempted already as a way to mitigate some of the metadata performance issues and while there is a marked reduction in metadata overhead using block devices and sparse files, it's more of an admission that for the time being, MFS use cases may simply be outside the scope for certain HPC workloads.  It's not entirely a big deal because I've ran into similar problems with GlusterFS and discovered there really is only two open-source file systems that are capable of dealing with this problem at scale. Lustre and Ceph, and both suck from a setup and management perspective.  I digress. Thanks for all the help you're able to provide.  I don't want my laziness to impede or encroach upon a stable system for the sake of convenience.
